modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerMetadataTest.java

```diff
@@ -105,6 +108,96 @@ public void testMetadataList() {
             assertContains(log, out, "typeName=Type_" + typeNum);
     }
 
+    /**
+     * Checks that command remove when provided with absent type name prints informative message and
+     * returns corresponding error code (invalid argument).
+     *
+     * Test scenario:
+     * <ol>
+     *     <li>
+     *         Execute meta --remove command passing non-existing type name.
+     *     </li>
+     *     <li>
+     *         Verify that invalid argument error code is returned, message is printed to command output.
+     *     </li>
+     *     <li>
+     *         Execute meta --remove command passing non-existing type id.
+     *     </li>
+     *     <li>
+     *         Verify that invalid argument error code is returned, message is printed to command output.
+     *     </li>
+     * </ol>
+     */
+    @Test
+    public void testMetadataRemoveWrongType() {
+        injectTestSystemOut();
+
+        String wrongTypeName = "Type01";
+
+        assertEquals(EXIT_CODE_INVALID_ARGUMENTS, execute("--meta", "remove", "--typeName", wrongTypeName));
+
+        assertContains(log, testOut.toString(), "Failed to remove binary type, type not found: " + wrongTypeName);
+
+        int wrongTypeId = 42;
+
+        assertEquals(EXIT_CODE_INVALID_ARGUMENTS, execute("--meta", "remove", "--typeId", Integer.toString(wrongTypeId)));
+
+        assertContains(log, testOut.toString(), "Failed to remove binary type, type not found: ");
+        assertContains(log, testOut.toString(), "0x" +
+            Integer.toHexString(wrongTypeId).toUpperCase() + " (" + wrongTypeId + ")");
+    }
+
+    /**
+     * Checks that commands --meta list and --meta remove don't cause registering system or internal classes
+     * in binary metadata.
+     *
+     * Test scenario:
+     * <ol>
+     *     <li>
+     *         Put a value of a user class to the default cache that requires registering new binary type for that class.
+     *     </li>
+     *     <li>
+     *         Run --meta list command to check that class is registered.
+     *     </li>
+     *     <li>
+     *         Destroy the cache. Remove binary type with --meta remove command.
+     *     </li>
+     *     <li>
+     *         Run --meta list again to check that no additional system or internal classes were registered.
+     *     </li>
+     * </ol>
+     */
+    @Test
+    public void testMetadataForInternalClassesIsNotRegistered() {
+        injectTestSystemOut();
+
+        IgniteCache<Object, Object> dfltCache = grid(0).getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        dfltCache.put(1, new TestValue());
+
+        Collection<BinaryType> metadata = crd.context().cacheObjects().metadata();
+
+        assertEquals(EXIT_CODE_OK, execute("--meta", "list"));
+
+        assertEquals(metadata.toString(), 1, metadata.size());
+
+        assertContains(log, testOut.toString(), "typeName=" + TestValue.class.getTypeName());
+
+        grid(0).destroyCache(DEFAULT_CACHE_NAME);
```

Comment:
I think it makes sense to add one more assert here that metadata is not empty yet after cache destroy(assertEquals(metadata.toString(), 1, metadata.size());)

Reply:
Good point, I modified test as suggested and confirmed locally that it passes with the change.